zaza/model.py

```diff
@@ -1320,12 +1330,30 @@ async def async_wait_for_application_states(model_name=None, states=None,
             timed_out = int(time.time() - start) > timeout
             issues = []
             for application in applications_left.copy():
+                check_info = states.get(application, {})
                 app_data = model.applications.get(application, None)
                 units = list(app_data.units)
-                # if there are no units the application is 'ready'
-                if len(units) > 0:
-                    check_info = states.get(application, {})
 
+                # all_okay is a Boolean of the current state.  It starts as
+                # True, but if False by the end of the checks, then the
+                # application is not ready.
+                all_okay = True
+
+                # if there are no units then the application may not be ready.
+                # However, if the caller explicitly allows that situation then
+                # we gate on that.
+                num_expected = check_info.get('num-expected-units', None)
+                if num_expected is not None:
+                    if len(units) != num_expected:
+                        all_okay = False
+                else:
+                    # num_expected is None, so 0 units means we are still
+                    # waiting
+                    if len(units) == 0:
+                        continue
```

Comment:
`continue` is correct, but in order to remain in the spirit of the rest of the function, this should be `all_okay = False` instead, which will then eventually lead to `continue` on line 1407. What do you think?

Reply:
Yes, I think this is a reasonable suggestion as it does tie in with the comment.  I'm toying with the idea of backing out the all_okay flag at this point and simplifying the logic slightly, which I will do next.  Thanks for the review comment :)